bip-0360.mediawiki

```diff
@@ -176,137 +177,196 @@ It is proposed to use SegWit version 3. This results in addresses that start wit
+Additional follow-on BIPs will be needed to implement signature aggregation, and full BIP-32 compatibility
 (if possible) <ref name="bip-32">BIP-32 relies on elliptic curve operations to derive keys from xpubs to support
 watch-only wallets, which PQC schemes may not support.</ref>. However, until specialized quantum cryptography hardware
-is widespread and signature aggregation schemes are thoroughly vetted, P2QRH addresses should be an adequate
-intermediate solution that provides meaningful protection against quantum threats.
+is widespread and signature aggregation schemes are thoroughly vetted, P2QRH addresses are an intermediate solution 
+to quantum threats.
 
 == Specification ==
 
 We define the signature scheme and transaction structure as follows.
 
-=== Descriptor Format ===
+=== Pay to Quantum Resistant Hash (P2QRH) ===
 
-To integrate P2QRH into existing wallet software and scripts, we introduce a new output descriptor function
-<code>qrh()</code>. This function represents a P2QRH output, similar to how <code>wpkh()</code> and <code>tr()</code>
-are used for P2WPKH and P2TR outputs, respectively.
+A P2QRH output is simply the root of the tapleaf merkle tree defined as in [https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki BIP-341] 
+and used as an internal value in P2TR.
 
-The <code>qrh()</code> function takes a threshold value and multiple key specifications grouped by key type. The format is:
+To construct a P2QRH output we follow the same process as [https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki BIP-341]
+to compute the tapscript merkle root. However, instead of the root of the Merkle tree being hashed together with the internal
+key in P2QRH the root is hashed by itself using the tag "QuantumRoot".
 
-  qrh(threshold, keytype(0x01, [hash1, hash2, ...]), keytype(0x02, [hash1, hash2, ...]), ...)
+<source>
+D = tagged_hash("TapLeaf", bytes([leaf_version]) + ser_script(script))
+CD = tagged_hash("TapBranch", C + D)
+CDE = tagged_hash("TapBranch", E + CD)
+ABCDE = tagged_hash("TapBranch", AB + CDE)
+Root = tagged_hash("QuantumRoot", ABCDE)
+</source>
 
-Where:
+TODO: Add image of P2QRH output merkle tree structure.
+
+Validation follows the pattern of a P2TR script spend but without the P2TR public key or internal key.
+A P2QRH input provides the following:
+
+<source>
+script witness stack,
+tapleaf script,
+control block = [version byte, m*32 byte merkle path] # m is the depth of the merkle tree
+</source>
+
+The script witness stack provides the same functionality as it does in P2TR. It places input values on the stack to 
+be evaluated by the tapscript. 
+
+A P2QRH script witness stack differs in only one way from a P2TR script witness stack. P2QRH does not limit the size 
+of the witness stack elements to 520 bytes. This is needed because PQ signatures and public keys can be larger than 
+520 bytes.
```

Comment:
Should we just increase the limit to the largest signature size (SLH-DSA) ?

Reply:
That's a good idea.